libdnf-plugins/actions/actions.cpp

```diff
@@ -0,0 +1,725 @@
+                    break;
+                case 'r':
+                    str[dst_pos++] = '\r';
+                    break;
+                case 't':
+                    str[dst_pos++] = '\t';
+                    break;
+                case 'v':
+                    str[dst_pos++] = '\v';
+                    break;
+                default:
+                    str[dst_pos++] = str[src_pos];
+            }
+            escape = false;
+        } else if (str[src_pos] == '\\') {
+            escape = true;
+        } else {
+            str[dst_pos++] = str[src_pos];
+        }
+    }
+    str.resize(dst_pos);
+}
+
+void Actions::setup(const std::vector<Action> & trans_actions) {
+    if (trans_actions.empty()) {
+        return;
+    }
+
+    std::vector<CommandToRun> commands_to_run;      // std::vector because we want to keep the order of commands
+    std::set<CommandToRun> unique_commands_to_run;  // std::set is used to detect duplicate commands
+
+    for (const auto & action : trans_actions) {
+        if (auto [substituted_args, subst_error] = substitute_args(nullptr, nullptr, action); !subst_error) {
+            for (auto & arg : substituted_args) {
+                unescape(arg);
+            }
+            CommandToRun cmd_to_run{action.command, std::move(substituted_args)};
+            if (auto [it, inserted] = unique_commands_to_run.insert(cmd_to_run); inserted) {
+                commands_to_run.push_back(std::move(cmd_to_run));
+            }
+        }
+    }
+
+    // execute commands
+    for (auto & cmd : commands_to_run) {
+        execute_command(cmd);
+    }
+}
+
+void Actions::parse_action_files() {
+    const auto & config = base.get_config();
+    const auto & conf_dirs = config.pluginconfpath().get_value();
+    for (const auto & dir : conf_dirs) {
+        auto action_dir_path = std::filesystem::path(dir) / "actions.d";
+        std::vector<std::filesystem::path> action_paths;
+        std::error_code ec;  // Do not report errors if config_dir_path refers to a non-existing file or not a directory
+        for (const auto & p : std::filesystem::directory_iterator(action_dir_path, ec)) {
+            if ((p.is_regular_file() || p.is_symlink()) && p.path().extension() == ".action") {
+                action_paths.emplace_back(p.path());
+            }
+        }
+        std::sort(action_paths.begin(), action_paths.end());
+
+        for (const auto & path : action_paths) {
+            std::ifstream action_file(path);
+            std::string line;
+            int line_number = 0;
+            while (std::getline(action_file, line)) {
+                ++line_number;
+                if (line.empty() || line[0] == '#') {
+                    continue;
+                }
+                auto pkg_filter_pos = line.find(':');
+                if (pkg_filter_pos == std::string::npos) {
+                    throw ActionsPluginError(M_("Error in file \"{}\" at line {}"), path.native(), line_number);
```

Comment:
This error and the two identical throws below are very generic. While it may be "obvious" the colons are missing, we can be more specific in the message, e.g. say ""Error in file \"{}\" on line {}: \\"HOOK:PKG_SPEC:DIRECTION:CMD\\" format expected". The names of the items are just as an example, I'm not sure these already have better established names in the old plugin or in the docs.